### PR TITLE
CR-1126396 xbtop output does not look correct

### PR DESCRIPTION
**Pull request by dbenusov-xilinx**
#### Problem solved by the commit
https://jira.xilinx.com/browse/CR-1126396
When xbtop is running for a u55c board, the output looks keep rolling and most data is not visible

#### Bug / issue (if any) fixed, which PR introduced the bug, how it was discovered
Discovered on cards that have a large memory bank to display after running a validation test

#### How problem was solved, alternative solutions (if any) and why they were rejected
The problem was solved by splitting up the data displays onto multiple pages the users can go through. The memory report was consolidated into two internal reports rather than three.

#### Risks (if any) associated the changes in the commit
Risk of data not being presented properly using xbtop if an unexpected case appears. Possible crashes in these scenarios as well. Resizing the window does not dynamically reprint data, the program must be stopped and restarted.

#### What has been tested and how, request additional testing if necessary
Remory report pages:
image
image
image
Resized terminal window:
image
Report with single page:
image
Report with no xclbin loaded:
image
image
image
Updated memory utilization to have percentages:
image

#### Documentation impact (if any)
None, unless we would like to document the paging feature.

### CODE REVIEW

**Comment by dbenusov-xilinx** [comment]
> Looks good to me. Can you test when no xclbin is loaded? just a sanity check Also, should "Utilization" have "%" in it?

Good suggestions! I added a test for no xclbin and found a bug present in xbtop which I fixed in here. Also added the '%' character into the utilization printout

File: src/runtime_src/core/common/info_memory.cpp [hunk 1]
```diff
@@ -468,7 +468,7 @@ scheduler_update_stat(const xrt_core::device *device)
     // lock xclbin
     std::string xclbin_uuid = xrt_core::device_query<xq::xclbin_uuid>(dev);
     // dont open a context if xclbin_uuid is empty
-    if(xclbin_uuid.empty())
+    if(xq::xclbin_uuid::is_empty(xclbin_uuid))
       return;
```
Comment (stsoe) [inline src/runtime_src/core/common/info_memory.cpp:472]: Space between `if` and `(` always please.

Do just
```
if (xclbin_uuid.empty() || !xrt::uuid(xclbin_uuid))
  return;
```

File: src/runtime_src/core/common/query_requests.h [hunk 2]
```diff
@@ -698,6 +699,12 @@ struct xclbin_uuid : request
 
   virtual boost::any
   get(const device*) const = 0;
+  
+  static bool
+  is_empty(const std::string& uuid) 
+  {
+    return (boost::equals(uuid, "00000000-0000-0000-0000-000000000000") || boost::equals(uuid, ""));
+  }
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:707]: Remove

**Review by stsoe — CHANGES_REQUESTED** [review]

**Review by rozumx — CHANGES_REQUESTED** [review]
Please schedule a demo.

File: src/runtime_src/core/common/query_requests.h (hunk 2 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/query_requests.h:707]: And here Shivangi and I though we invented the wheel! Cool feature with the boolean overload on the uuid :)

File: src/runtime_src/core/common/query_requests.h [hunk 3]
```diff
@@ -29,6 +29,7 @@
 #include <stdexcept>
 #include <boost/any.hpp>
 #include <boost/format.hpp>
+#include <boost/algorithm/string.hpp>
```
Comment (stsoe) [inline src/runtime_src/core/common/query_requests.h:32]: Is this used?

**Review by stsoe — CHANGES_REQUESTED** [review]

**Review by stsoe — APPROVED** [review]
Python is beyond me and you got rid of all C++ changes :-)

**Review by AShivangi — APPROVED** [review]

File: src/runtime_src/core/common/info_memory.cpp [hunk 4]
```diff
@@ -467,8 +467,8 @@ scheduler_update_stat(const xrt_core::device *device)
   try {
     // lock xclbin
     std::string xclbin_uuid = xrt_core::device_query<xq::xclbin_uuid>(dev);
-    // dont open a context if xclbin_uuid is empty
-    if(xclbin_uuid.empty())
+    // dont open a context if xclbin_uuid is empty or is all zeros
+    if (xclbin_uuid.empty() || !xrt::uuid(xclbin_uuid))
```
Comment (rozumx) [inline src/runtime_src/core/common/info_memory.cpp:471]: What was this change required?

File: src/runtime_src/core/tools/xbtop/ReportMemory.py [hunk 5]
```diff
@@ -28,108 +29,186 @@ def update(self, dev):
...
+            XBUtil.print_warning(term, lock, start_y + offset, "Data unavailable. Acceleration image not loaded")
+            return offset + 1
+
+        for i in range(self.report_length):
+            # The current element to be parsed depends on what page has been requested
+            index = i + (page * self.report_length)
+            # Ensure that our index does not exceed the input data size. This may happen on the last page
+            if (index < len(memories)):
+                memory = memories[index]
+                size = int(memory['range_bytes'], 16)
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbtop/ReportMemory.py:72]: Magic number

File: src/runtime_src/core/tools/xbtop/ReportMemory.py [hunk 6]
```diff
@@ -28,108 +29,186 @@ def update(self, dev):
...
+            if (index < len(memories)):
+                line.append(str(index))
+                memory = memories[index]
+                line.append(memory['tag'])
+                line.append(memory['type'])
+                line.append(memory['extended_info']['temperature_C'] if 'temperature_C' in memory['extended_info'] else "--")
+                line.append(XBUtil.convert_size(int(memory['range_bytes'],16)))
+                line.append(XBUtil.convert_size(int(memory['extended_info']['usage']['buffer_objects_count'],0)))
+                
+                size = int(memory['range_bytes'], 16)
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbtop/ReportMemory.py:124]: Magic number

File: src/runtime_src/core/tools/xbtop/ReportMemory.py [hunk 7]
```diff
@@ -28,108 +29,186 @@ def update(self, dev):
...
-            line.append(XBUtil.convert_size(int(dma_metrics[i]['card_to_host_bytes'], 16)))
-            data.append(line)
+            # The current element to be parsed depends on what page has been requested
+            index = i + (page * self.report_length)
+            # Ensure that our index does not exceed the input data size. This may happen on the last page
+            if (index < len(dma_metrics)):
+                dma_metric = dma_metrics[index]
+                line.append(str(i))
+                line.append(dma_metric['channel_id'])
+                line.append(XBUtil.convert_size(int(dma_metric['host_to_card_bytes'], 16)))
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbtop/ReportMemory.py:178]: Magic number

File: src/runtime_src/core/tools/xbtop/xbtop.py [hunk 8]
```diff
@@ -66,7 +67,8 @@ def running_counter(term, lock, len_x):
 # Key table
 def print_footer(term, lock, y_len):
     footer_buf = [
-        XBUtil.fg.yellow + "'n'/'p' - Next/Previous Report" + XBUtil.fg.reset,
+        XBUtil.fg.yellow + "'p'/'n' - Previous/Next Report" + XBUtil.fg.reset,
```
Comment (rozumx) [inline src/runtime_src/core/tools/xbtop/xbtop.py:70]: Thank you for swapping the 'p' and 'n'.  It helps keep the menu and the key description consistent.

**Review by rozumx — APPROVED** [review]
Overall looks good, though it could some some additional comments.

File: src/runtime_src/core/common/info_memory.cpp (hunk 4 above)
Comment (dbenusov-xilinx) [reply src/runtime_src/core/common/info_memory.cpp:471]: There was an bug on the dynamic report page that would cause a crash due to the xclbin uuid being all 0s. Since the uuid is not "empty" the software in C++ and Python would attempt to open context on a non existent file which caused strange issues. C++ it would crash when opening the context, in Python is would crash when closing which failed to catch for one reason or another.
